### DISCUSSION TITLE
Qualities upgrade for movies

### DISCUSSION THREAD

**Discussion by whiteout12** [Flexget/Flexget#3732] [Q&A]
I'm using the upgrade qualities for my series filters where it grabs the better quality if available later on. Can this be used for movies as well?

This feature for series is very useful as I do not have to have a separate upgrade-task. 

As an example is this possible?
Mymovietask:
  qualities:
    - 2160p remux dolbyvision 
    - 2160p remux
    - 2160p webdl
    - 1080p remux

**Comment by girtsl** [comment]
Can you paste your full movie task? 
You should be able to achieve what you want with `quality`, `upgrade` (to track downloaded qualities), and disabling `seen` - to not reject the downloaded titles right away.  
Also make sure you don't remove the entry on a successful match - which is probably the default action for many download tasks. The caveat here is that you'll need some condition to remove the title from input list after your desired quality has been reached - this is where `if` might help.
All this considered, your single movie task will become quite complicated and going with a separate upgrade task might be preferable.

**Reply by whiteout12** [reply]
I already have a separate upgrade task for movies so I am solving this but it's just so much more complicated than the qualities solution for series.

So that's why I thought a discussion on a proposal of porting the functionality of qualities to work outside the series plugin could be started. 

Would it be handy to have this ?

**Reply by girtsl** [reply]
I bet the maintainers would be happy to get a PR implementing the quality upgrade for movies and many users would appreciate the reduced complexity as well.

**Comment by gazpachoking** [comment]
Have you tried the [upgrade](https://flexget.com/en/Plugins/upgrade) plugin? It isn't quite as easy as with series, but that's what it's designed for.

**Reply by whiteout12** [reply]
No have not tried it because I don't think it will work for my needs. It's based on the rank of the qualities in the quality plugin right?

I want to be able to skip some qualities in between the final target and the lowest quality accepted. In my example above I just want to be able to go for the qualities I mentioned and not pick everything above 1080p for example.

So the the qualities plug-in for series is superior to the upgrade plugin.

Maybe I have understood the upgrade plugin wrong and would be happy if I could get some help setting it up with an example
matching my example above.

**Reply by gazpachoking** [reply]
You'd also use the quality plugin to only allow the acceptable quality levels.
```yaml
# other plugins
quality:
- 720p-1080p webrip
- 1080p webdl
- 2180p bluray
upgrade:
  target: 2180p
  timeframe: 1 week
```

**Reply by whiteout12** [reply]
Aha ok so, I can have multiple entries in the quality plugin and the upgrade plugin will keep track of what has been downloaded? How are these ranked among each other?

Is "2160p Webdl" higher than "1080 Blu-ray remux" for example?

**Reply by gazpachoking** [reply]
It still isn't quite as tight as the integration in the series plugin. The quality plugin can just restrict what is acceptable, the upgrade plugin only looks at things that pass that filter, and will get something that's better until the 'target' value is hit. Quality is sorted like in [this table](https://flexget.com/Plugins/quality#qualities). The leftmost column has the highest priority (such that 2160p is always considered better than 1080p, no matter the other attributes) and the next column to the right is only considered if there is a tie (such that 1080p bluray is better than 1080p hdtv).

**Reply by whiteout12** [reply]
Ok good, because I didn't know it was a left to right priority among the columns. Then maybe I can get this to work after all.

I will play around with it. Thanks for the help.

**Reply by whiteout12** [reply]
Coming back here as I have found an issue that I didn't think of.

First I didn't get the upgrade to accept other (better) qualities that didn't meet the actual target. So I thought that I needed to use "on_lower" to accept everything lower than taget (but still better than what was already downloaded.

So that worked (I think), but then is seems that "seen_movies" plugin is overriding and rejected because same title has already been downloaded.

I then disabled "seen_movies" which let me download but then it accepted lower qualities than what was already downloaded :(

So can this be handled in any way?
